eth/tracers/taiko_api.go

```diff
@@ -0,0 +1,449 @@
+// - config: The configuration for the trace.
+//
+// Returns:
+// - *rpc.Subscription: A subscription to the trace results.
+// - error: An error if the operation fails, or if the end block is not after the start block.
+//
+// Note:
+//   - The function requires a notifier to be present in the context, otherwise it
+//     returns an error indicating that notifications are unsupported.
+func (api *API) ProvingPreflights(ctx context.Context, start, end rpc.BlockNumber, config *TraceConfig) (*rpc.Subscription, error) {
+	if start == 0 {
+		return nil, fmt.Errorf("start block must be greater than 0")
+	}
+	// Adjust the start block to the parent one
+	start -= 1
+
+	from, err := api.blockByNumber(ctx, start)
+	if err != nil {
+		return nil, err
+	}
+	to, err := api.blockByNumber(ctx, end)
+	if err != nil {
+		return nil, err
+	}
+	if from.Number().Cmp(to.Number()) >= 0 {
+		return nil, fmt.Errorf("end block (#%d) needs to come after start block (#%d)", end, start)
+	}
+	// Tracing a chain is a **long** operation, only do with subscriptions
+	notifier, supported := rpc.NotifierFromContext(ctx)
+	if !supported {
+		return &rpc.Subscription{}, rpc.ErrNotificationsUnsupported
+	}
+	sub := notifier.CreateSubscription()
+
+	resCh := api.provingPreflights(from, to, config, sub.Err())
+	go func() {
+		for result := range resCh {
+			notifier.Notify(sub.ID, result)
+		}
+	}()
+	return sub, nil
+}
+
+// provingPreflights traces the execution of blocks from `start` to `end` and returns a channel
+// that streams the results of the preflight checks for each block. The function uses multiple
+// goroutines to parallelize the tracing process.
+//
+// Parameters:
+// - start: The starting block for tracing.
+// - end: The ending block for tracing.
+// - config: Configuration for tracing, including the tracer to use and reexec settings.
+// - closed: A channel to signal when tracing should be aborted.
+//
+// Returns:
+// - A channel that streams the results of the preflight checks for each block.
+//
+// The function performs the following steps:
+// 1. Determines the number of blocks to trace and the number of threads to use.
+// 2. Initializes the tracing configuration and state tracker.
+// 3. Starts multiple goroutines to trace the blocks concurrently.
+// 4. Feeds the blocks into the tracers and processes the results.
+// 5. Streams the results back to the caller through the returned channel.
+//
+// The tracing process involves fetching the blocks, preparing the state, and tracing each
+// transaction within the blocks. The results include the transaction trace results, account
+// proofs, contract codes, and ancestor hashes.
+func (api *API) provingPreflights(start, end *types.Block, config *TraceConfig, closed <-chan error) chan *provingPreflightResult {
+	reexec := defaultTraceReexec
+	if config != nil && config.Reexec != nil {
+		reexec = *config.Reexec
+	}
+	blocks := int(end.NumberU64() - start.NumberU64())
```

Comment:
Currently, the start must be >= the end, so is it a bug?

Reply:
Assert in method `ProvingPreflights` above